### ISSUE TITLE
CA2000 - False positives

### ISSUE THREAD

**Issue by MrMikeJJ** [dotnet/roslyn-analyzers#1404]
#### Analyzer package
Microsoft.CodeQuality.Analyzers

#### Analyzer
CA2000: Dispose Objects Before Losing Scope

#### Repro steps

1. Follow https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2000-dispose-objects-before-losing-scope for example good code
2a) Allow IDE0017 to be applied - Object initialization can be simplified - on the temporary object
or
2b) Add a catch statement
or
2c) Encapsulate the return object (unsure if this is a false positive or a real positive

#### Expected behavior
CA2000 not to be reported

#### Actual behavior
CA2000 is reported

Sample code file attached.
[CA2000.txt](https://github.com/dotnet/roslyn-analyzers/files/1502312/CA2000.txt)

Since i think it is 2 (potentially 3) false positives, but all on the same rule, i didn't know if you would want me to report them grouped together, or as a separate report for each. Sorry if i shouldn't have grouped them.

thanks
Mike

(And thanks for Roslyn and the Analyzers - they are great :))

**Comment by mavasani** [comment]
@MrMikeJJ Can you please tell us the version of the analyzer package and VS version you are using?

**Comment by MrMikeJJ** [comment]
Hi Mavasani,
Certainly :)
I am using Visual Studio Professional 2017 - Version 15.4.4
This is found using the analyzer built into Visual Studio, under "Run Code Analysis on Solution".
According to the output window, this is C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\Team Tools\Static Analysis Tools\FxCop\FxCopCmd.exe
That program reports itself as 
Microsoft (R) FxCop Command-Line Tool, Version 15.0 (15.0.26906.1) X86
If that is what you were after ?
thanks
Mike

**Comment by mavasani** [comment]
@MrMikeJJ That is the legacy version of FxCop tool which performs binary analysis, not the FxCopAnalyzers package. CA2000 has not yet been implemented, see [this issue](https://github.com/dotnet/roslyn-analyzers/issues/476) and current [stub implementation](https://github.com/dotnet/roslyn-analyzers/blob/6ff9c17d8fc05fae8b1c48b7a4d25b1e080f0b97/src/Microsoft.CodeQuality.Analyzers/Core/QualityGuidelines/DisposeObjectsBeforeLosingScope.cs#L34). I am closing this issue as a duplicate of https://github.com/dotnet/roslyn-analyzers/issues/476, we hope to implement the rule soon and will address this report at that time.

**Comment by mavasani** [comment]
Re-activating this to validate the repro is fixed and a regression test added by the CA2000 implementation in https://github.com/dotnet/roslyn-analyzers/pull/1583. We should not regress this with #1580

**Comment by drewnoakes** [comment]
This scenario gives me a false positive with 2.6.x

```vb.net
Dim disposeMe As IDisposable = Nothing
Try
    For Each c In "Foo"
        If disposeMe Is Nothing Then
            disposeMe = New MemoryStream()
            '           ~~~~~~~~~~~~~~~~~~
        End If
    Next
Finally
    If disposeMe IsNot Nothing Then
        disposeMe.Dispose()
    End If
End Try
```

**Comment by mavasani** [comment]
Thanks @drewnoakes - I just confirmed this also repros with 2.9.x package, I will investigate.

**Comment by mavasani** [comment]
Both cases fixed with https://github.com/dotnet/roslyn-analyzers/pull/2146#merged-event in 2.9.x package.

**Comment by amrmahdi** [comment]
@mavasani I'm using version 2.9.0-beta1.19158.2 and I'm getting a false positive for the following code 

```cs
            try
            {
                using (var ms = new MemoryStream())
                {
                    // do something with memory stream
                }
            }
            catch
            {
            }
```

**Comment by drewnoakes** [comment]
@amrmahdi can you create a new issue please?

**Comment by amrmahdi** [comment]
> @amrmahdi can you create a new issue please?

Created #2201
